Libraries/LibC/stdio.cpp

```diff
@@ -911,25 +910,53 @@ void perror(const char* s)
 static int parse_mode(const char* mode)
 {
     int flags = 0;
-    // NOTE: rt is a non-standard mode which opens a file for read, explicitly
-    // specifying that it's a text file
-    if (!strcmp(mode, "r") || !strcmp(mode, "rb") || !strcmp(mode, "rt"))
+    switch (*mode) {
+    case 'r':
```

Comment:
@bugaevc a little late for that lol

Reply:
What do you mean?

Reply:
Just that it already does the parsing thing you suggested above.

Reply:
But it parses the first character different to others, while I'm suggesting it could not do that and treat all the characters the same way.

Reply:
Ah, I see, I managed to misread that somehow.
I could do that, but I believe it's done this way to prevent bogus modes such as "rw" or "awwwwwww"

Reply:
I don't insist on doing it that way, it just sounds simpler and more generic to me.

Reply:
I'm not against it, I'm just saying we explicitly define its semantics first if we do that,
i.e. what should "rw" mean? same as "r+"? Repetition can be ignored, conflicting information...not so much

Reply:
> we explicitly define its semantics first if we do that

We don't have to do that, either. Garbage in, garbage out — you'll probably get some nonsensical combination of flags and the kernel will reject it with `EINVAL`. That being said, "rw" will naturally resolve to `O_RDONLY | O_WRONLY`, which on Serenity is the same as `O_RDWR`, so it'll behave as expected.

Reply:
Garbage in, garbage out, but at which level? Why should an abstraction leak failures it can detect?
Ideally, fopen wouldn't do a syscall if you feed it crap at all

Reply:
To spell it out: I'm not suggesting we should officially support and define semantics for any combination of any mode characters in any order. I'm mainly suggesting we could simplify the implementation, and as a side effect enable specifying characters in any order. It still means that if you pass an invalid combination, you're going to get yourself in trouble, just perhaps into some different sort of trouble.

Reply:
And I'm saying that we shouldn't allow things that we have no defined semantics for.
If we don't support crap like "awr", we shouldn't let it get down to the kernel before it gets possibly rejected (although I don't think there's any particular combination that would fail, since "awr" is just "a+" with our flags).

But why leak crap when we can simply not?

Reply:
Dr. POSIX says, "The mode argument points to a string. If the string is one of the following, the file shall be opened in the indicated mode. Otherwise, the behavior is undefined."

Feeding a possibility nonsensical combination of flags to the kernel in those cases sounds alright to me.

> Why should an abstraction leak failures it can detect?

If you want to, you can detect those cases, e.g.
```c++
case 'w':
    ASSERT(!(mode & O_RDONLY));
    mode |= O_WRONLY;
```
I'm fine with both doing and not doing this.

Reply:
Well...that's a whole lot of asserts for a mere "er" support, I'm not convinced that the trade-off is worth the extra weirdness.
But stdio is practically yours, so I'll go ahead and do it your way.

Reply:
> If we don't support crap like "awr", we shouldn't let it get down to the kernel before it gets possibly rejected (although I don't think there's any particular combination that would fail, since "awr" is just "a+" with our flags).

> But why leak crap when we can simply not?

To me this sounds like asking for `O_WRONLY | O_APPEND | O_CREAT | O_WRONLY | O_TRUNC | O_CREAT | O_RDONLY`, just specified as a string. That combination doesn't make sense, but why not pass it to the kernel and let it run its logic to decide whether it makes sense or not.

On the other hand it's `fopen()` that does the translation, so it might as well run some validation. But it's not "leaking crap" if it doesn't — it would just pass to the kernel the exact flags it has been asked to.

Reply:
`fopen` is already being biased with regards to its input by ignoring things it doesn't understand, but alright, let's leave the does-this-flag-make-sense to the kernel.

Reply:
Speaking of modes, where does this `rt` mode come from?
text would be the default mode anyway, and we don't even do anything with it

Reply:
No idea. It's not in POSIX ('b' is).

It was added in https://github.com/SerenityOS/serenity/commit/c2a8c4cedd8c682fb861b446d4bf1f151e97ee2d

Reply:
Sounds like some program trying to be fancy, but ok.